### ISSUE TITLE
Option to change button order

### ISSUE THREAD

**Issue by ianwitherow** [makeusabrew/bootbox#291]
It'd be useful to have the ability to move the buttons around. As it is now, for bootbox.confirm the OK button is on the right and Cancel is on the left, however in all major browsers, OK is on the left and Cancel is on the right.

**Comment by burmester** [comment]
This is not enough.. 
If you want the button in the center of the modal.. Then the modal-footer needs a text-algin:center..

**Comment by tarlepp** [comment]
@burmester you could always use plain bootstrap dialog and add your own CSS styles to that.

**Comment by MrCroft** [comment]
@tarlepp it would have been much more simple if the buttons would just be placed in the order they're found in the "buttons" option

**Comment by tarlepp** [comment]
That will need some work to get working. Eg. we need to change that object structure to be an array, which contains each button object, etc.

But please make a PR about this issue.

**Comment by Real-Gecko** [comment]
+1
Would be really nice to have this feature.
Currently I noticed that prompt and confirm dialogs have success on the right, while dialogs have success on the left. Really confusing.

**Comment by dutchiexl** [comment]
+1

**Comment by dzaeny** [comment]
+1

**Comment by pablonautic** [comment]
+1

**Comment by vdusa** [comment]
+1

**Comment by tarlepp** [comment]
Anyone has a PR for this?

**Comment by jedd-ahyoung** [comment]
Is there any reason that the `buttons` property is an object and not an array? Are the keys actually used for anything?

**Comment by tarlepp** [comment]
Yeah those buttons should be an array, no arguee there. So once again I rise question: Anyone has a PR for this?

**Comment by tiesont** [comment]
@tarlepp I'll see if I can come up with something.

**Comment by makeusabrew** [comment]
Found this from reviewing #534 which linked #525 which led to here.

There was a very good reason why buttons switched from an array to an object, but it's lost to the depths of my memory for now. I'll rack my brains and try and remember why I changed it in the first place. Re-opening just for visibility's sake.

**Comment by makeusabrew** [comment]
I _think_ the main reason it changed was as per my comment in #431; the primary use case for the buttons object is to use the label -> callback shorthand:

```
buttons: {
  "label one": function() {
    // callback code when button clicked
  },
  "label two": function() {
  }
}
```

That said, it changed [over three years ago](https://github.com/makeusabrew/bootbox/commit/02a695b0e3f9226fa3cf55360f607c705112a9a6) and my memory isn't what it used to be, so I can't justify all the reasons I had for doing it at the time. I can see from that commit that they keys are certainly at least used internally in favour of array indices.

Nevertheless, I've yet to hear a convincing argument for changing it back (particularly because it would be a massive breaking change), and I don't really buy the argument for the key order not being guaranteed because I've never, _ever_ seen the order change when iterating an object literal's keys in today's browsers. Programatically slicing and dicing an object, sure, that's more difficult, but not something I want to encourage anyway (with bootbox).

Happy to be enlightened though :). @tarlepp @tiesont  @jedd-ahyoung et al?

**Comment by nsarafornccn** [comment]
Can anyone please help me with the position of buttons on bootbox dialog?

Sample Code:
 bootbox.dialog({
                            message: ""
                            buttons: {
                                success: {
                                    label: "Add Note",
                                    className: "btn-success"
                                },
                                danger: {
                                    label: "Close",
                                    className: "btn-red"
                                 }

I'm not sure what text I need to add to ClassName to make the buttons appear on top instead of bottom and there is a position property which I tried adding with "top" value but it didn't work.

Thanks!

**Comment by tarlepp** [comment]
@nsarafornccn does Bootstrap support that ?

**Comment by nsarafornccn** [comment]
No idea tarlepp. Just looking around If there is any possibility. I might have to choose a different approach if there is no such option.

**Comment by tarlepp** [comment]
@nsarafornccn seems like that is not supported http://getbootstrap.com/javascript/#modals

**Comment by nsarafornccn** [comment]
Np. Thanks for the quick response! @tarlepp

**Comment by rusanov-vladimir** [comment]
+1. I haven't found any way to swap position of Confirm and Cancel buttons for Confirm dialog. I was expecting buttons to be created in the same order they are declared, so if I declare them like:

	buttons: {
		confirm: {
			label: 'Yes',
			className: 'btn-primary'
		},
		cancel: {
			label: 'No',
			className: 'btn-default'
		}
	}

I was hoping to see 'Yes' to the left and 'No' to the right.

**Comment by tarlepp** [comment]
@rusanov-vladimir have you tried to add ```className: "pull-right"``` to both of your your buttons, and then swap those places if buttons was wrong order ?

**Comment by tarlepp** [comment]
@rusanov-vladimir you mean like this? http://plnkr.co/edit/9xxeoBxZU602gs5Oqsnh?p=preview

**Comment by rusanov-vladimir** [comment]
@tarlepp Thank you for a quick response! Yes, I've read about this workaround, but when using `pull-right` the margin between buttons disappears and it's starting to look not so bootstrap-like... I would like to be able to keep the layout as it is, but just switching buttons.
P.S. I even tried a workaround that manipulates the DOM to switch the order of buttons(http://jsfiddle.net/9L3A9/embedded/result/), it worked but caused problems for bootbox.alert and generally looks ugly...

**Comment by tarlepp** [comment]
so for that margin problem just make your own class eg. ```my-pull-right``` which adds that margin, those are just plain CSS classes.

Basically use your own classes to get the desired output.

**Comment by rusanov-vladimir** [comment]
Well, the problem with that approach is that I got a standard css theme that is based on bootstrap, so I suppose not to write any custom css classes. The idea behind this is that changing the theme should affect all the applications and there would be no need to update every single css class according to the theme changes. Bootbox has no css, it's completely based on bootstrap and it fits perfectly, except for this buttons order scenario...

**Comment by tarlepp** [comment]
And that is not supported - sorry for that.

I was just telling how you could easily solve your problem with one CSS class - personally I don't see any harm on that. But if that is not for you maybe you should use pure Bootstrap and build that modal from scratch.

**Comment by tarlepp** [comment]
And maybe you could make some PR about that button order - so everyone could benefit that ?

**Comment by rusanov-vladimir** [comment]
Yeah, If there would be no way to solve the issue with workarounds I'll switch to pure bootstrap dialog. 
I've been reading other issues related to the button order and it seems like a there were already some attempts to implement it, which failed/got rejected. Do you have any thoughts about how to implement this?

**Comment by tiesont** [comment]
@rusanov-vladimir  Technically speaking, if you forked this repo and then merged https://github.com/makeusabrew/bootbox/pull/534, you could set the order of your buttons. You'd have to figure out how to migrate the init code yourself, though.

**Comment by rusanov-vladimir** [comment]
@tiesont Thank you for the link. As I understand that solution is a breaking change to the API and that's why it's not merged(and probably will not be merged). I would like to have solution integrated in the master branch, so it would be easy to maintain.
@tarlepp Regarding the PR, I've took a look in the code and came up with a simple solution that shouldn't be breaking. Just adding a new property to the `button` called `index`. Based on that Index to sort buttons and create them in the order according to the index. I've tried to implement it, but my JS knowledge is way out of date, so probably someone could suggest improvements to that [code](https://github.com/rusanov-vladimir/bootbox/commit/0c4509ea6272ffe5d4bbe3659f5df49043de760b). It works for my scenario, but I haven't tested it well.
Please, let me know if this solution seems rational and I should submit a PR.

**Comment by rusanov-vladimir** [comment]
Is there any change to get this fixed in master any time soon(this month)? Or at least to have things moving?

**Comment by tiesont** [comment]
Not likely. I'm not particularly for it, so unless @tarlepp wants to put the work in, I don't see this happening prior to a 5.0 release, _if_ the change actually gets made - @makeusabrew hasn't exactly endorsed moving in this direction.

If you _did_ come up with a PR that enabled button order without changing the current API, and includes tests to prove it, I'd be more likely to push for a merge.

**Comment by tarlepp** [comment]
And may I remind you, that button order can be changed via simple CSS until we have a real solution for this. So if you guys are hurry with this, I would just make couple of line CSS to make this happen.

Personally I don't have time atm to make this happen + changes that I think are breaking changes - so not so easy process.

**Comment by tiesont** [comment]
@rusanov-vladimir If button order is super important, you may want to consider another library. https://nakupanda.github.io/bootstrap3-dialog/ looks like it uses an array for buttons, so I would assume the order you declare them in is the order in which they're built.

**Comment by tarlepp** [comment]
I made simple plunker to demonstrate how easy is to make this happen http://plnkr.co/edit/NuO17C0jSevsQuSKwZat?p=preview with current release - I hope this solves this problem for now.

And if you're not happy with this, just make a PR as @tiesont said.

**Comment by rusanov-vladimir** [comment]
Thanks for the support guys, I'll consider the options.

**Comment by jonny64bit** [comment]
+1 would be good if this could be supported out of the box.

**Comment by KaduAmaral** [comment]
+1 this will be very helpful

**Comment by tiesont** [comment]
I don't plan on implementing this in the v4.x branch, but if you want to test the v5.x branch, this has been implemented as of #631 - see #545 for what else is changing. @tarlepp Unless you want to port this to v4, I will probably close this issue.

**Comment by WilliamCasasola** [comment]
I did this and work fine for me:

className: _className,
        buttons: {
            confirm: {
                label: 'Aceptar',
                className: 'btn-success myButtonClass'
            },
            cancel: {
                label: 'Cancelar',
                className: 'btn-danger pull-right '
            }
        },

Css:
.myButtonClass
{
  margin-right: 5px;  
}
![untitled](https://user-images.githubusercontent.com/18342280/32841391-cb37e1ec-c9df-11e7-9432-721ba5e3d8a3.png)
